### DISCUSSION TITLE
Implementing Sofort and EPS payment methods with Airwallex

### DISCUSSION THREAD

**Discussion by sudhirrnanda** [juspay/hyperswitch#4144] [General]
### Update on AirWallex Problem Statements
AirWallex mandates the use of an organization email for Sandbox account setup. As a result, we'd like to ask candidates to select alternative problem statements. We will be adding more problem statements soon. 

### Context 

Embracing diversity in payments is one of the core tenets of Hyperswitch. And we look forward to integrating a large variety of payment processors (connectors) and payment flows around the world.

Airwallex is a financial technology company that provides businesses with global payment solutions, including cross-border payments, foreign exchange, and risk management, with easy integration. In Hyperswitch parlance, Airwallex is one of the connectors we have integrated with. As of now, hyperswitch supports card payments and google pay wallet payments through Airwallex

We look forward to including some more critical payment methods through Airwallex

### Problem Statement 

Sofort and EPS are online banking based payment methods that allow end users to complete payments in their own banking environment. 

- Sofort is popular in Germany, Austria and Switzerland and quickly gaining popularity all over the EU

- EPS, while originally developed by Austrian Banks, is also a very popular payment method in the EU

**The goal is to extend hyperswitch’s integration with Airwallex needs to support Sofort and EPS**

### How to get started:

1. [Set up a Rust environment and run Hyperswitch locally](https://github.com/juspay/hyperswitch/blob/main/docs/try_local_system.md) 

2. [Signup and create a sandbox account on Airwallex](https://www.airwallex.com/) to get the API Credentials

### How to Implement:

Now you can get started with the implementation: 

1. Read and understand the [general guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md) on contributing to Hyperswitch

2. Import postman collection of hyperswitch from [here](https://www.postman.com/hyperswitch/workspace/hyperswitch-development/collection/25176162-630b5353-7002-44d1-8ba1-ead6c230f2e3) 

3. Understand the implementation of all different payment flows and [how to add a connector](https://github.com/juspay/hyperswitch/blob/main/add_connector.md) 

4. As mentioned earlier for Airwallex, card payments are implemented. Test out this flow from postman using these [test resources](https://www.airwallex.com/docs/payments__test-card-numbers) and understand the API implementation 

5. Study the relevant API Documentation of Airwallex to implement [Sofort](https://www.airwallex.com/docs/payments__eu-and-uk__sofort.__desktopmobile-website-browser) and [EPS](https://www.airwallex.com/docs/payments__eu-and-uk__eps__desktopmobile-website-browser) through Airwallex

6. Implement Sofort and EPS using the API references in point 4. Modify the following file to implement Sofort and EPS

> crates/router/src/connector/airwallex/transformers.rs
> crates/router/src/connector/airwallex.rs

7. Test the following flows - Payment, Capture, Void and Refund using postman for your  local implementation and generate relevant screenshots of the test cases.

### Expected Outcomes: 

Submit a pull request to the [juspay/hyperswitch](https://github.com/juspay/hyperswitch) repository including the relevant screenshots of the test cases. 

Brownie Points: Providing clear and comprehensive documentation and code comments.

Note: We welcome multiple PRs for this problem statement

### More useful Resources:
[Setting up Airwallex on Hyperswitch](https://docs.hyperswitch.io/hyperswitch-cloud/connectors/available-connectors/airwallex)

**Comment by sagnik26** [comment]
Hi I would like to take this one..

**Reply by neerajkumar-n** [reply]
@sagnik26 - you can get started with this, you can raise a PR once done. We will review this and ensure to fill the Google form.

You can learn more on Problem statement here: https://github.com/juspay/hyperswitch/wiki/Hyperswitch-Hiring-Challenge

**Comment by AnishKacham** [comment]
would love to have a go!

**Reply by manojradhakrishnan** [reply]
Please try creating a pull request for the feature.

**Comment by vasudevanubrolu** [comment]
Hi I would love to take this up

**Reply by manojradhakrishnan** [reply]
Please try creating a pull request for the feature.

**Comment by GodLJonathan** [comment]
Hi @c10n3, I wanted to work on the same problem. Is it first come, first served, or can multiple people work on the same problem? Please clarify. Thanks!

**Reply by manojradhakrishnan** [reply]
Multiple developers can attempt the same problem. We value approach and clean code more than anything else.

**Comment by manav-chan** [comment]
I am done with setting up the Rust development environment for Hyperswitch and tried the basic APIs, and now I am researching on this issue..

**Reply by manojradhakrishnan** [reply]
Good luck. Please try creating a pull request for the feature.

**Comment by CuriousCorrelation** [comment]
Hi @manojradhakrishnan just a heads up - discussion https://github.com/juspay/hyperswitch/discussions/4145 mentions Sofort and EPS, along with SEPA and BACS in point 6

> Implement SEPA and BACS using the API references in point 5. Modify the following file to implement Sofort and EPS

I covered both in my [PR](https://github.com/juspay/hyperswitch/pull/4222), let me know if I should include this discussion thread in PR description as well.

**Reply by neerajkumar-n** [reply]
@CuriousCorrelation - I see your PR, but i don't think you have filled in the google form. Why don't you fill the google form: https://forms.gle/QQfA5fdG37QvCDio8

**Reply by manojradhakrishnan** [reply]
Thanks for letting know. Fixed the description.

**Reply by CuriousCorrelation** [reply]
@neerajkumar-n Thanks you so much for the remainder! Just waiting to hear back from a couple of recent collaborations to make sure everything is up to date.

A quick question if you don't mind, it seems the form expects one challenge, tho will it be alright to include all https://github.com/juspay/hyperswitch/discussions/4179 at https://github.com/juspay/hyperswitch/issues/4191, https://github.com/juspay/hyperswitch/discussions/4144 + https://github.com/juspay/hyperswitch/discussions/4145 at https://github.com/juspay/hyperswitch/pull/4222, perhaps comma separated? I believe they might offer insight in both planning and execution, if it wouldn't be too much trouble of course.

**Reply by neerajkumar-n** [reply]
Yes @CuriousCorrelation, please do add them with a comma separated.

**Reply by CuriousCorrelation** [reply]
Hey @neerajkumar-n thanks for your patience, form's filled and sent! Email is same as my username here if you'd like to confirm. Thanks again for the remainder.

**Comment by tusharpandeyzee** [comment]
Hi @sudhirrnanda @neerajkumar-n, I have submitted PR https://github.com/juspay/hyperswitch/pull/4310 for implementing this. Request to review. Thanks.

**Reply by neerajkumar-n** [reply]
@tusharpandeyzee - we recommend you to go through this and fill in the google form:  https://github.com/juspay/hyperswitch/wiki/Hyperswitch-Hiring-Challenge

**Reply by tusharpandeyzee** [reply]
@neerajkumar-n i had filled the google form after raising the PR, waiting for a reply.
